repo2docker/buildpacks/base.py

```diff
@@ -271,16 +271,54 @@ def stencila_manifest_dir(self):
         # ${STENCILA_ARCHIVE_DIR}/${STENCILA_ARCHIVE}/manifest.xml
 
         self._stencila_manifest_dir = None
+
+        if os.path.exists('binder'):
+            return self._stencila_manifest_dir
+
         for root, dirs, files in os.walk("."):
             if "manifest.xml" in files:
+                self.log.debug("Found a manifest.xml at %s", root)
                 self._stencila_manifest_dir = root.split(os.path.sep, 1)[1]
                 self.log.info(
-                    "Found stencila manifest.xml in %s",
+                    "Using stencila manifest.xml in %s",
                     self._stencila_manifest_dir,
                 )
                 break
         return self._stencila_manifest_dir
 
+    @property
+    def stencila_manifest_contexts(self):
+        """Find the stencila manifest contexts if it exists"""
+        if hasattr(self, '_stencila_manifest_contexts'):
+            return self._stencila_manifest_contexts
+
+        # look at the content of the jats.xml file to extract the required execution contexts
+
+        self._stencila_manifest_contexts = None
+
+        if os.path.exists('binder'):
+            return self._stencila_manifest_contexts
+
+        for root, dirs, files in os.walk("."):
+            for filename in files:
+                if filename.endswith(".jats.xml"):
+                    self.log.debug("Found a .jats.xml: %s", filename)
+                    self._stencila_manifest_contexts = set()
+
+                    # extract code languages from file
+                    with open(os.path.join(root, filename)) as f:
+                        for line in f:
+                            match = re.match('.*language="(.+?)"', line)
```

Comment:
Can we use a XML parser to read in the jats.xml file and find these attributes? Using a regex on a XML file feels like we will get weird breakage in a few months or so.

Reply:
The regex does match, correct. It did not hurt so far, but you're right to keep things cleaner.

I just marked the PR as "WIP", because using an XML parser is no quick fix I'll do today.